R/epidatacall.R

```diff
@@ -68,187 +58,219 @@ create_epidata_call <- function(endpoint, params, meta = NULL,
   ))
 }
 
+#' Fetches the data
+#'
+#' `fetch` usually returns the data in tibble format, but a few the endpoints
+#' only support the JSON classic format. In that case a JSON-like nested list
+#' structure is returned instead.
+#'
+#' @rdname epidata_call
+#' @param epidata_call an instance of `epidata_call`
+#' @param fields filter fields
+#' @param disable_date_parsing disable automatic date parsing
+#' @return
+#' - For `fetch`: a tibble or a JSON-like list
+#' @export
+#'
+fetch <- function(epidata_call, fields = NULL, disable_date_parsing = FALSE) {
+  stopifnot(inherits(epidata_call, "epidata_call"))
+  stopifnot(is.null(fields) || is.character(fields))
+  stopifnot(is.logical(disable_date_parsing), length(disable_date_parsing) == 1)
+
+  if (epidata_call$only_supports_classic) {
+    return(fetch_classic(epidata_call, fields))
+  } else {
+    return(fetch_tbl(epidata_call, fields, disable_date_parsing))
+  }
+}
+
 #' Fetches the data and returns a tibble
 #' @rdname epidata_call
 #'
 #' @param epidata_call an instance of `epidata_call`
 #' @param fields filter fields
 #' @param disable_date_parsing disable automatic date parsing
+#' @param method the method to use to fetch the data
 #' @importFrom readr read_csv
 #' @importFrom httr stop_for_status content
 #' @importFrom rlang abort
 #' @return
 #' - For `fetch_tbl`: a [`tibble::tibble`]
+#' @importFrom tibble as_tibble
 #'
 #' @export
 fetch_tbl <- function(epidata_call, fields = NULL, disable_date_parsing = FALSE) {
   stopifnot(inherits(epidata_call, "epidata_call"))
   stopifnot(is.null(fields) || is.character(fields))
   stopifnot(is.logical(disable_date_parsing), length(disable_date_parsing) == 1)
+
   if (epidata_call$only_supports_classic) {
-    rlang::abort("the endpoint only supports the classic message format, due to a non-standard behavior",
+    rlang::abort("This endpoint only supports the classic message format, due to a non-standard behavior. Use fetch_classic instead.",
       epidata_call = epidata_call,
       class = "only_supports_classic_format"
     )
   }
-  r <- fetch_csv(epidata_call, fields)
-  meta <- epidata_call$meta
-  fields_pred <- fields_to_predicate(fields)
-  col_names <- c()
-  col_types <- list()
-  for (i in seq_len(length(meta))) {
-    info <- meta[[i]]
-    if (fields_pred(info$name)) {
-      col_names <- c(col_names, info$name)
-      col_types[info$name] <- info_to_type(info, disable_date_parsing)
-    }
-  }
-  tbl <- if (length(col_names) > 0) {
-    readr::read_csv(r, col_types = col_types)
-  } else {
-    readr::read_csv(r)
-  }
-
 
-  if (!disable_date_parsing) {
-    # parse weeks
-    columns <- colnames(tbl)
-    for (i in seq_len(length(meta))) {
-      info <- meta[[i]]
-      if (info$name %in% columns && info$type == "epiweek") {
-        tbl[[info$name]] <- parse_api_week(tbl[[info$name]])
-      }
-    }
-  }
-  tbl
+  df <- fetch_classic(epidata_call, fields, disable_data_frame_parsing = FALSE)$epidata
```

Comment:
todo: if we don't `rlang::warn` on non-"success" message in `fetch_classic`, it must be done here.

Reply:
(and I'd also like to raise an error on truncated requests even if we decide against it in `fetch_classic`.  there's no real way to make use of the results you do get.)

Reply:
changed to raise on non-success in fetch_classic and return result$epidata